crates/typst/src/eval/mod.rs

```diff
@@ -1076,23 +1077,71 @@ impl Eval for ast::Binary<'_> {
     }
 }
 
-/// Apply a basic binary operation.
+impl<'a> Eval for ast::ChainedComp<'a> {
+    type Output = Value;
+
+    #[tracing::instrument(name = "ChainedComp::eval", skip_all)]
+    fn eval(self, vm: &mut Vm) -> SourceResult<Self::Output> {
+        let mut side = |lhs: &dyn Fn() -> ast::Expr<'a>,
+                        op,
+                        rhs: &dyn Fn() -> ast::Expr<'a>| match op {
```

Comment:
I think this is a case where dyn Fn could actually be worse than generics. besides the match could just map to `ops::..` instead of the whole apply thing. in general, I feel like the evaluation part could be simplified a bit.